src/github.com/getlantern/flashlight/ui/service.go

```diff
@@ -89,14 +114,29 @@ func read() {
 		}
 
 		// Delegating response to the service that registered with the given type.
-		if services[env.Type] != nil {
-			// Pass this message and continue reading another one.
-			go func() {
-				services[env.Type].In <- b
-			}()
-		} else {
-			log.Errorf("Message type %s belongs to an unkown service.", env.Type)
+		if services[envType.Type] == nil {
+			log.Errorf("Message type %v belongs to an unkown service.", envType.Type)
+			return
 		}
 
+		env := &envelope{}
+		err = json.Unmarshal(b, env)
+		if err != nil {
+			log.Errorf("Unable to unmarshal message of type %v: %v", envType.Type, err)
+			return
+		}
+		// Pass this message and continue reading another one.
+		services[env.Type].in <- env.Message
```

Comment:
Note - I'm no longer doing this in a goroutine.  The problem with spawning goroutines is that if the client started sending tons of messages, we'd end up piling up a bunch of goroutines.

A goroutine isn't necessary here anyway since the channel is buffered.